File: Exercises/Exercise 1 (Markus)/shared variable/c/foo.c
// Compile with `gcc foo.c -Wall -std=gnu99 -lpthread`, or use the makefile
// The executable will be named `foo` if you use the makefile, or `a.out` if you use gcc directly

#include <pthread.h>
#include <stdio.h>

int i = 0;
pthread_mutex_t mutex;

void* incrementing() {

    // Add mutex to avoid race conditions

    for (int j = 0; j < 1000000; j++) {
        pthread_mutex_lock(&mutex);
        i++;
        pthread_mutex_unlock(&mutex);
    }

    return NULL;
}

void* decrementing() { 
    
    //Add mutex to avoid race conditions

    for (int j = 0; j < 1000001; j++) {
        pthread_mutex_lock(&mutex);
        i--;
        pthread_mutex_unlock(&mutex);
    }

    return NULL;
}

int main() {
    // Initialize the two threads
    pthread_t incrementingThread, decrementingThread; 

    // Initialize the mutex
    pthread_mutex_init(&mutex, NULL); 

    pthread_create(&incrementingThread, NULL, incrementing, NULL);
    pthread_create(&decrementingThread, NULL, decrementing, NULL);

    // Makes sure that the threads finish before proceeding
    pthread_join(incrementingThread, NULL); 
    pthread_join(decrementingThread, NULL);

    //Clean up the mutex
    pthread_mutex_destroy(&mutex);

    printf("The magic number is: %d\n", i);
    return 0;
}

File: Exercises/Exercise 1 (Markus)/shared variable/go/foo.go
// Use `go run foo.go` to run your program

package main

import (
    . "fmt"
    "runtime"
    "sync"
    "time"
)

var (
    i     = 0
    mutex sync.Mutex //Initialize the mutex
)

func incrementing() {

    // Add mutex to avoid race conditions

    for j := 0; j < 1000000; j++ {
        mutex.Lock()
        i++
        mutex.Unlock()
    }
}

func decrementing() {

    // Add mutex to avoid race conditions

    for j := 0; j < 1000001; j++ {
        mutex.Lock()
        i--
        mutex.Unlock()
    }
}

func main() {

    // Allow go runtime to use up to two threads for executing go-routines
    runtime.GOMAXPROCS(2)


    // Executing the go-routines
    go incrementing()
    go decrementing()

    //Pauses the main-function to give the go-routines enough time to finish
    time.Sleep(500 * time.Millisecond)
    Println("The magic number is:", i)
}